mongo/integration/gridfs_spec_test.go

```diff
@@ -0,0 +1,498 @@
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package integration
+
+import (
+	"bytes"
+	"encoding/hex"
+	"io/ioutil"
+	"path"
+	"testing"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/bsontype"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/internal/testutil/assert"
+	"go.mongodb.org/mongo-driver/mongo/gridfs"
+	"go.mongodb.org/mongo-driver/mongo/integration/mtest"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/x/bsonx"
+	"go.mongodb.org/mongo-driver/x/bsonx/bsoncore"
+)
+
+type gfsTestFile struct {
+	Data  gfsData   `bson:"data"`
+	Tests []gfsTest `bson:"tests"`
+}
+
+type gfsData struct {
+	Files  []bson.Raw  `bson:"files"`
+	Chunks []bsonx.Doc `bson:"chunks"`
+}
+
+type gfsTest struct {
+	Description string     `bson:"description"`
+	Arrange     gfsArrange `bson:"arrange"`
+	Act         gfsAct     `bson:"act"`
+	Assert      gfsAssert  `bson:"assert"`
+}
+
+type gfsArrange struct {
+	Data []bsonx.Doc `bson:"data"`
+}
+
+type gfsAct struct {
+	Operation string   `bson:"operation"`
+	Arguments bson.Raw `bson:"arguments"`
+}
+
+type gfsAssert struct {
+	Result interface{} `bson:"result"`
+	Error  string      `bson:"error"`
+	Data   []bsonx.Doc `bson:"data"`
+}
+
+const (
+	gfsTestsDir       = "../../data/gridfs"
+	gfsFiles          = "fs.files"
+	gfsChunks         = "fs.chunks"
+	gfsExpectedFiles  = "expected.files"
+	gfsExpectedChunks = "expected.chunks"
+)
+
+var (
+	gfsDeadline    = time.Now().Add(time.Hour)
+	downloadBuffer = make([]byte, 100)
+)
+
+func TestGridFSSpec(t *testing.T) {
+	mt := mtest.New(t, noClientOpts)
+	defer mt.Close()
+
+	for _, file := range jsonFilesInDir(mt, gfsTestsDir) {
+		mt.Run(file, func(mt *mtest.T) {
+			runGfsTestFile(mt, path.Join(gfsTestsDir, file))
+		})
+	}
+}
+
+func runGfsTestFile(mt *mtest.T, filePath string) {
+	content, err := ioutil.ReadFile(filePath)
+	assert.Nil(mt, err, "ReadFile error for %v: %v", filePath, err)
+
+	var testFile gfsTestFile
+	err = bson.UnmarshalExtJSONWithRegistry(specTestRegistry, content, false, &testFile)
+	assert.Nil(mt, err, "UnmarshalExtJSONWithRegistry error: %v", err)
+
+	for _, test := range testFile.Tests {
+		mt.Run(test.Description, func(mt *mtest.T) {
+			runGfsTest(mt, test, testFile)
+		})
+	}
+}
+
+func runGfsTest(mt *mtest.T, test gfsTest, testFile gfsTestFile) {
+	chunkSize := setupGfsTest(mt, testFile.Data)
+	if chunkSize == 0 {
+		chunkSize = gridfs.DefaultChunkSize
+	}
+	bucket, err := gridfs.NewBucket(mt.DB, options.GridFSBucket().SetChunkSizeBytes(chunkSize))
+	assert.Nil(mt, err, "NewBucket error: %v", err)
+	err = bucket.SetWriteDeadline(gfsDeadline)
+	assert.Nil(mt, err, "SetWriteDeadline error: %v", err)
+	err = bucket.SetReadDeadline(gfsDeadline)
+	assert.Nil(mt, err, "SetReadDeadline error: %v", err)
+
+	arrangeGfsCollections(mt, test.Arrange)
```

Comment:
The previous runner ran the `arrange` section in the individual operation functions (e.g `runDownloadTest`). The new runner always runs the section before executing any operations.